### PR TITLE
HWKAPM-775 Only define 'client' prefix on endpoints where fragment on…

### PR DESCRIPTION
**Pull request by objectiser**
…ly contains single top level Producer node, to distinguish it from the endpoint used for the server.

### CODE REVIEW

**Comment by objectiser** [comment]
@pavolloffay Could you review please?

Reviewer notes:
* Simplified approach for identifying the endpoint reference associated with initial trace fragments that don't begin with a Consumer.
* Previously the endpoint ref was simply prefixed with a 'client:' and the URI/op of the first Producer node was used. This was for two reasons, (1) to distinguish the endpoint ref from the one used by the receiving server's fragment, and (2) because top level Component nodes captured by the instrumenting agent approach may not have useful URI/Op for use in this endpoint.
* Now focusing on opentracing approach, where more meaningful data could be provided, the only situation that needs to be treated in a special way is an initial trace fragment that just has a single Producer node. In this case, we still need to prefix the endpoint reference to distinguish it from the server.

**Comment by objectiser** [comment]
retest this please

File: api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java [hunk 1]
```diff
@@ -96,6 +96,8 @@ public void testDecodeNonClientURI() {
     @Test
     public void testGetSourceEndpointConsumer() {
         Trace trace = new Trace();
+        trace.setTraceId("1");
```
Comment (pavolloffay) [inline api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java]: How is this change related to this PR?

File: api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java [hunk 2]
```diff
@@ -110,6 +112,8 @@ public void testGetSourceEndpointConsumer() {
     @Test
     public void testGetSourceEndpointClientProducer() {
         Trace trace = new Trace();
+        trace.setTraceId("1");
```
Comment (pavolloffay) [inline api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java]: How is this change related to this PR?

File: api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java [hunk 3]
```diff
@@ -122,39 +126,25 @@ public void testGetSourceEndpointClientProducer() {
     }
 
     @Test
-    public void testGetSourceEndpointClientComponent() {
+    public void testGetSourceEndpointNotClientComponentProducer() {
+        // As this initial fragment does not have a single Producer node,
+        // its endpoint ref will be based on the component's URI and operation
         Trace trace = new Trace();
+        trace.setTraceId("1");
```
Comment (pavolloffay) [inline api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java]: How is this change related to this PR?

File: server/api/src/test/java/org/hawkular/apm/server/api/utils/SourceInfoUtilTest.java [hunk 4]
```diff
@@ -53,7 +53,8 @@
     @Test
     public void testInitialiseIds() throws RetryAttemptException {
         Trace trace = new Trace();
-        trace.setFragmentId("trace1");
+        trace.setTraceId("trace1");
```
Comment (pavolloffay) [inline server/api/src/test/java/org/hawkular/apm/server/api/utils/SourceInfoUtilTest.java]: How is this change related to this PR?

File: server/processors/src/test/java/org/hawkular/apm/server/processor/tracecompletiontime/TraceCompletionInformationInitiatorTest.java [hunk 5]
```diff
@@ -95,6 +95,29 @@ public void testProcessSingleInitialFragmentConsumer() throws RetryAttemptExcept
...
+        p.setUri("uri");
+        p.addInteractionCorrelationId("pid");
+
+        trace.getNodes().add(p);
+
+        TraceCompletionInformationInitiator initiator = new TraceCompletionInformationInitiator();
+
+        TraceCompletionInformation ci = initiator.processOneToOne(null, trace);
+
+        assertNotNull(ci);
```
Comment (pavolloffay) [inline server/processors/src/test/java/org/hawkular/apm/server/processor/tracecompletiontime/TraceCompletionInformationInitiatorTest.java]: This is redundant.

File: api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java (hunk 1 above)
Comment (objectiser) [inline api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java]: The decision regarding whether to add the 'client:' prefix has to test whether the trace fragment is the initial (root) fragment, which is achieved by checking that the fragment and trace ids are the same. This required these tests to be updated to set the trace id, which was previously not required.

File: api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java (hunk 2 above)
Comment (objectiser) [inline api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java]: As above.

File: server/api/src/test/java/org/hawkular/apm/server/api/utils/SourceInfoUtilTest.java (hunk 4 above)
Comment (objectiser) [inline server/api/src/test/java/org/hawkular/apm/server/api/utils/SourceInfoUtilTest.java]: As above :)

File: server/processors/src/test/java/org/hawkular/apm/server/processor/tracecompletiontime/TraceCompletionInformationInitiatorTest.java (hunk 5 above)
Comment (objectiser) [inline server/processors/src/test/java/org/hawkular/apm/server/processor/tracecompletiontime/TraceCompletionInformationInitiatorTest.java]: Its redundant with knowledge of the current implementation - but in terms of the API the `processOneToOne` method does not guarantee to return a value.

**Comment by objectiser** [comment]
@pavolloffay Tests updated to remove unnecessary properties. Ready for re-review.

File: api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java (hunk 1 above)
Comment (pavolloffay) [inline api/src/test/java/org/hawkular/apm/api/utils/EndpointUtilTest.java]: I see otherwise it would fail on NPE `Trace#getinitialFragment()` :+1: . This reminds me if it is mentioned gitbook#writing OT provider that initial fragment should have the same id as the trace.

**Comment by objectiser** [comment]
@pavolloffay Thanks - and I'll make the update to the docs.
